hibernate-core/src/main/java/org/hibernate/engine/internal/AbstractEntityEntry.java

```diff
@@ -357,49 +357,54 @@ public boolean requiresDirtyCheck(Object entity) {
 	}
 
 	private boolean isUnequivocallyNonDirty(Object entity) {
-		if ( isSelfDirtinessTracker( entity ) ) {
-			final boolean uninitializedProxy;
-			if ( isPersistentAttributeInterceptable( entity ) ) {
-				final PersistentAttributeInterceptor interceptor =
-						asPersistentAttributeInterceptable( entity ).$$_hibernate_getInterceptor();
-				if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
-					EnhancementAsProxyLazinessInterceptor enhancementAsProxyLazinessInterceptor =
-							(EnhancementAsProxyLazinessInterceptor) interceptor;
-					return !enhancementAsProxyLazinessInterceptor.hasWrittenFieldNames(); //EARLY EXIT!
-				}
-				else {
-					uninitializedProxy = false;
-				}
+		return isSelfDirtinessTracker( entity )
+				? isNonDirtyViaTracker( entity )
+				: isNonDirtyViaCustomStrategy( entity );
+	}
+
+	private boolean isNonDirtyViaCustomStrategy(Object entity) {
+		if ( isPersistentAttributeInterceptable( entity ) ) {
+			final PersistentAttributeInterceptor interceptor =
+					asPersistentAttributeInterceptable( entity ).$$_hibernate_getInterceptor();
+			if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
+				// we never have to check an uninitialized proxy
+				// TODO: why do we not check !lazinessInterceptor.hasWrittenFieldNames()
+				//       as we do below in isNonDirtyViaTracker() ?
```

Comment:
@dreab8 This is a question for you.

Reply:
@gavinking  with dirty checking  setting a property will not trigger the initialization, while without dirty checking setting a property triggers the enhamced proxy initilaization and then the interceptor would not be of type `EnhancementAsProxyLazinessInterceptor`

Reply:
basically `writtenFieldNames` is populated only when dirty checking is enabled

Reply:
@dreab8 I guess I don't understand what is the difference between the two branches. They both do:

```java
if ( interceptor instanceof EnhancementAsProxyLazinessInterceptor ) {
```

and then one of them does the additional check, and one doesn't.

Reply:
@gavinking one branch deals wit the case dirty checking is not enabled, in this case it's enough `interceptor instanceof EnhancementAsProxyLazinessInterceptor` to be sure it's not dirty, the other branch deals with the case dirty checking is enabled and if `interceptor instanceof EnhancementAsProxyLazinessInterceptor` we need also to be sure it has not written field names to say that it's not dirty